README.md
[![Build Status](https://api.travis-ci.org/shurik2k5/yii2-upload-behavior.svg?branch=master)](https://travis-ci.org/shurik2k5/yii2-upload-behavior)
[![Total Downloads](https://img.shields.io/packagist/dt/shurik2k5/yii2-upload-behavior.svg)](https://packagist.org/packages/shurik2k5/yii2-upload-behavior)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/shurik2k5/yii2-upload-behavior/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/shurik2k5/yii2-upload-behavior/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/shurik2k5/yii2-upload-behavior/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/shurik2k5/yii2-upload-behavior/?branch=master)

Upload behavior for Yii 2
===========================

This behavior automatically uploads file and fills the specified attribute with a value of the name of the uploaded file.

In this behaviour added ability to load file from URL and local files, and attach behavior for two or more attributes.

This is an enhanced version of https://github.com/mohorev/yii2-upload-behavior/

Installation
------------

The preferred way to install this extension via [composer](http://getcomposer.org/download/).

Either run

```
php composer.phar require --prefer-dist shurik2k5/yii2-upload-behavior "*"
```

or add this code line to the `require` section of your `composer.json` file:

```json
"shurik2k5/yii2-upload-behavior": "*"
```

Usage
-----

### Upload file from input forms

Attach the behavior in your model:

```php
class Document extends ActiveRecord
{
    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            ['file', 'file', 'extensions' => 'doc, docx, pdf', 'on' => ['insert', 'update']],
        ];
    }

    /**
     * @return \yii\db\ActiveQuery
     */
    public function getCategory()
    {
        return $this->hasOne(Category::class, [ 'id' => 'id_category' ]);
    }

    /**
     * @inheritdoc
     */
    function behaviors()
    {
        return [
            [
                'class' => \mohorev\file\UploadBehavior::class,
                'attribute' => 'file',
                'scenarios' => ['insert', 'update'],
                'path' => '@webroot/upload/docs/{category.id}',
                'url' => '@web/upload/docs/{category.id}',
            ],
        ];
    }
}
```

Set model scenario in controller action:

```php
class Controller extends Controller
{
    public function actionCreate($id)
    {
        $model = $this->findModel($id);
        $model->setScenario('insert'); // Note! Set upload behavior scenario.
        
        ...
        ...
    }
}

```

Example view file:

```php
<?php $form = ActiveForm::begin(['options' => ['enctype' => 'multipart/form-data']]); ?>
    <?= $form->field($model, 'image')->fileInput() ?>
    <div class="form-group">
        <?= Html::submitButton('Submit', ['class' => 'btn btn-primary']) ?>
    </div>
<?php ActiveForm::end(); ?>
```
### Upload file from url
To load file from url
```php
$model = new Document();
$model->setScenario('update'); //Use scenarion for validation and load file
$model->uploadFromUrl('file', 'https://i.ytimg.com/vi/yfJbKba5xYM/maxresdefault.jpg');
$model->save();
```
### Upload file from file
To load file from local file
```php
$model = new Document();
$model->setScenario('update'); //Use scenarion for validation and load file
$model->uploadFromFile('file', \Yii::getAlias('@webroot/images/02.jpg'));
$model->save();
```

### Get path in you application 
Get upload url
```php
$model->getUploadUrl('file');
```

Get upload path
```php
$model->getUploadPath('file');
```

### Upload image and create thumbnails

Thumbnails processing requires [yiisoft/yii2-imagine](https://github.com/yiisoft/yii2-imagine) to be installed.

Attach the behavior in your model:

```php
class User extends ActiveRecord
{
    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            ['image', 'image', 'extensions' => 'jpg, jpeg, gif, png', 'on' => ['insert', 'update']],
        ];
    }

    /**
     * @inheritdoc
     */
    public function behaviors()
    {
        return [
            [
                'class' => \mohorev\file\UploadImageBehavior::class,
                'attribute' => 'image',
                'scenarios' => ['insert', 'update'],
                'placeholder' => '@app/modules/user/assets/images/userpic.jpg',
                'path' => '@webroot/upload/user/{id}',
                'url' => '@web/upload/user/{id}',
                //if need create all thumbs profiles on image upload
                'createThumbsOnSave' => true,
                //if need create thumb for one profile only on request by getThumbUploadUrl() method
                'createThumbsOnRequest' => true,
                //if you want to remove original upload file after images thumbs was generated
                'deleteOriginalFile' => true,
                'thumbs' => [
                    'thumb' => ['width' => 400, 'quality' => 90],
                    'preview' => ['width' => 200, 'height' => 200],
                    'news_thumb' => ['width' => 200, 'height' => 200, 'bg_color' => '000'],
                ],
            ],
        ];
    }
}
```

### Remove upload image and all thumbs
```php
//remove image and all thumbs from 'photo' attribute
$model->deleteImage('photo');

```

## Flexible configuration for path and URL generation 

More flexible configuration for `path` and/or `url` behavior properties is that to use 
callbacks or array for defining `path` or `url` generation logic.

#### I. Via callbacks

```php
/**
 * @inheritdoc
 */
public function behaviors()
{
    return [
        [
            'class' => \mohorev\file\UploadImageBehavior::class,
            'attribute' => 'image',
            'scenarios' => ['insert', 'update'],
            'placeholder' => '@app/modules/user/assets/images/userpic.jpg',
            'path' => function ($model) {
                /** @var \app\models\UserProfile $model */
                $basePath = '@webroot/upload/profiles/';
                $path = implode('/', array_slice(str_split(md5($model->id), 2), 0, 2));
                return $basePath . $path;
            },
            'url' => function ($model) {
                /** @var \app\models\UserProfile $model */
                $baseUrl = '@web/upload/profiles/';
                $path = implode('/', array_slice(str_split(md5($model->id), 2), 0, 2));
                return $baseUrl . $path;
            },
            'thumbs' => [
                'thumb' => ['width' => 400, 'quality' => 90],
                'preview' => ['width' => 200, 'height' => 200],
                'news_thumb' => ['width' => 200, 'height' => 200, 'bg_color' => '000'],
            ],
        ],
    ];
}
```

#### II. Via array configuration by defining class and its static methods for path/URL generation

```php
/**
 * @inheritdoc
 */
public function behaviors()
{
    return [
        [
            'class' => \mohorev\file\UploadImageBehavior::class,
            'attribute' => 'image',
            'scenarios' => ['insert', 'update'],
            'placeholder' => '@app/modules/user/assets/images/userpic.jpg',
            'path' => [UserProfile::class, 'buildAvatarPath'],
            'url' => [UserProfile::class, 'buildAvatarUrl'],
            'thumbs' => [
                'thumb' => ['width' => 400, 'quality' => 90],
                'preview' => ['width' => 200, 'height' => 200],
                'news_thumb' => ['width' => 200, 'height' => 200, 'bg_color' => '000'],
            ],
        ],
    ];
}

/**
 * Define two static methos in your model for path and URL generation
 */ 
/**
 * @param \app\models\UserProfile|\yii\db\ActiveRecord $profile
 * @return string
 */
public static function buildAvatarPath(UserProfile $model)
{
    $basePath = '@webroot/upload/profiles/';
    $path = implode('/', array_slice(str_split(md5($model->id), 2), 0, 2));

    return $basePath . $path;
}

/**
 * @param \app\models\UserProfile|\yii\db\ActiveRecord $profile
 * @return string
 */
public static function buildAvatarUrl(UserProfile $model)
{
    $baseUrl = '@web/upload/profiles/';
    $path = implode('/', array_slice(str_split(md5($model->id), 2), 0, 2));

    return $baseUrl . $path;
}
```  

Example view file:

```php
<?php $form = ActiveForm::begin(['options' => ['enctype' => 'multipart/form-data']]); ?>
    <div class="form-group">
        <div class="row">
            <div class="col-lg-6">
                <!-- Original image -->
                <?= Html::img($model->getUploadUrl('image'), ['class' => 'img-thumbnail']) ?>
            </div>
            <div class="col-lg-4">
                <!-- Thumb 1 (thumb profile) -->
                <?= Html::img($model->getThumbUploadUrl('image'), ['class' => 'img-thumbnail']) ?>
            </div>
            <div class="col-lg-2">
                <!-- Thumb 2 (preview profile) -->
                <?= Html::img($model->getThumbUploadUrl('image', 'preview'), ['class' => 'img-thumbnail']) ?>
            </div>
        </div>
    </div>
    <?= $form->field($model, 'image')->fileInput(['accept' => 'image/*']) ?>
    <div class="form-group">
        <?= Html::submitButton('Submit', ['class' => 'btn btn-primary']) ?>
    </div>
<?php ActiveForm::end(); ?>
```

Behavior Options
-------

* attribute - The attribute which holds the attachment
* scenarios - The scenarios in which the behavior will be triggered
* instanceByName - Getting file instance by name, If you use UploadBehavior in `RESTfull` application and you do not need a prefix of the model name, set the property `instanceByName = false`, default value is `false`
* path - the base path or path alias to the directory in which to save files.
* url - the base URL or path alias for this file
* generateNewName - Set true or anonymous function takes the old filename and returns a new name, default value is `true`
* unlinkOnSave - If `true` current attribute file will be deleted, default value is `true`
* unlinkOnDelete - If `true` current attribute file will be deleted after model deletion.
* deleteEmptyDir - If `true` the **empty** directory will be deleted after model deletion, default value is `true`.

UploadImageBehavior additional Options
-------

* createThumbsOnSave - If `true` create all thumbs profiles on image upload
* createThumbsOnRequest - If `true` create thumb only for profile request by `getThumbUploadUrl('attribute', 'profile_name)` method. If `true` recommend to set `createThumbsOnSave` to `false`
* deleteOriginalFile - If `true` the **original upload image** will be deleted after images thumbs was generated, default value is `false`.\
**Attention** don't use with **createThumbsOnRequest** options, because thumbs generate on request (NOT after upload image) and after first profile thumb generation original file will be deleted!

### Attention!

It is prefered to use immutable placeholder in `url` and `path` options, other words try don't use related attributes that can be changed. There's bad practice. For example:

```php
class Track extends ActiveRecord
{
    public function getArtist()
    {
        return $this->hasOne(Artist::class, [ 'id' => 'id_artist' ]);
    }

    public function behaviors()
    {
        return [
            [
                'class' => \mohorev\file\UploadBehavior::class,
                'attribute' => 'image',
                'scenarios' => ['default'],
                'path' => '@webroot/uploads/{artist.slug}',
                'url' => '@web/uploads/{artist.slug}',
            ],
        ];
    }
}
```

If related model attribute `slug` will change, you must change folders' names too, otherwise behavior will works not correctly.

Authors
-----
* Alexander Mohorev <dev.mohorev@gmail.com>
* Alexey Samoylov <alexey.samoylov@gmail.com>
* Aleksandr Pogorelov <shurik2k5@gmail.com>


CHANGELOG.md
Yii2 upload behaviour change Log
===================================

1.2, December 09, 2020
---------------------
- Feat: Add public method to delete image and all thumbs (shurik2k5)

1.1, January 31, 2020
---------------------
- Fix: Ability attach behavior for two or more attributes (shurik2k5) 


tests/models/Gallery.php
<?php

namespace tests\models;

use mohorev\file\UploadImageBehavior;
use yii\db\ActiveRecord;

/**
 * Class User
 */
class Gallery extends ActiveRecord
{
    /**
     * @inheritdoc
     */
    public static function tableName()
    {
        return 'gallery';
    }

    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            [['image1', 'image2'], 'image', 'extensions' => 'jpg, jpeg, gif, png', 'on' => ['insert', 'update']],
        ];
    }

    /**
     * @inheritdoc
     */
    public function behaviors()
    {
        return [
            'image1' => [
                'class' => UploadImageBehavior::class,
                'attribute' => 'image1',
                'scenarios' => ['insert', 'update'],
                'path' => '@webroot/upload/gallery/{id}',
                'url' => '@web/upload/gallery/{id}',
                'placeholder' => '@tests/data/test-image.jpg',
                'createThumbsOnRequest' => true,
                'createThumbsOnSave' => false,
                'generateNewName' => false,
                'thumbs' => [
                    'thumb' => ['width' => 400, 'quality' => 90],
                    'preview' => ['width' => 200, 'height' => 200],
                ],
            ],
            'image2' => [
                'class' => UploadImageBehavior::class,
                'attribute' => 'image2',
                'scenarios' => ['insert', 'update'],
                'path' => '@webroot/upload/gallery/{id}',
                'url' => '@web/upload/gallery/{id}',
                'placeholder' => '@tests/data/test-image.jpg',
                'createThumbsOnRequest' => true,
                'createThumbsOnSave' => true,
                'generateNewName' => false,
                'thumbs' => [
                    'thumb' => ['width' => 400, 'quality' => 90],
                    'preview' => ['width' => 200, 'height' => 200],
                ],
            ],
        ];
    }
}


tests/UploadImageBehaviorTest.php
<?php

namespace tests;

use tests\models\Gallery;
use tests\models\User;
use yii\web\UploadedFile;

class UploadImageBehaviorTest extends DatabaseTestCase
{
    public function testFindUsers()
    {
        $data = User::find()->asArray()->all();
        $this->assertEquals(require(__DIR__ . '/data/test-find-users.php'), $data);
    }

    public function testFindUser()
    {
        $user = User::findOne(1);
        $this->assertEquals('admin', $user->nickname);
        $this->assertEquals('image-1.jpg', $user->image);
    }

    public function testGetFileInstance()
    {
        $file = UploadedFile::getInstanceByName('User[image]');
        $this->assertTrue($file instanceof UploadedFile);
    }

    public function testCreateUser()
    {
        $user = new User([
            'nickname' => 'Alex',
        ]);
        $user->setScenario('insert');

        $this->assertTrue($user->save());

        $path = $user->getUploadPath('image');
        $this->assertTrue(is_file($path));
        $this->assertEquals(sha1_file($path), sha1_file(__DIR__ . '/data/test-image.jpg'));
    }

    public function testResizeUser()
    {
        $user = User::findOne(1);
        $user->setScenario('update');

        $this->assertTrue($user->save());

        //request image by url
        $thumbUrl = $user->getThumbUploadUrl('image', 'thumb');
        $thumbPath = $user->getThumbUploadPath('image', 'thumb');

        $thumbInfo = getimagesize($thumbPath);
        $this->assertEquals(400, $thumbInfo[0]);
        $this->assertEquals(300, $thumbInfo[1]);

        //request image by url
        $previewUrl = $user->getThumbUploadUrl('image', 'preview');
        $previewPath = $user->getThumbUploadPath('image', 'preview');
        $previewInfo = getimagesize($previewPath);
        $this->assertEquals(200, $previewInfo[0]);
        $this->assertEquals(200, $previewInfo[1]);
    }

    public function testCreateUserImagesByRequestUrl()
    {
        $user = new User([
            'nickname' => 'Alex',
            'id' => 4,
        ]);
        $user->setScenario('insert');

        $this->assertTrue($user->save());

        $this->assertFalse(file_exists(__DIR__ . '/upload/user/' . $user->id . '/preview-test-image.jpg'));
        //request image by url
        $previewPath = $user->getThumbUploadUrl('image', 'preview');
        $this->assertTrue(file_exists(__DIR__ . '/upload/user/' . $user->id . '/preview-test-image.jpg'));
    }

    public function testGetPlaceholderOnNotExistAttribute() {
        $user = User::findOne(1);

        $image = $user->getThumbUploadUrl('not_exist_attribute', 'thumb');
        $this->assertContains('thumb-test-image.jpg', $image);
    }

    public function testFindGalleries()
    {
        $data = Gallery::find()->asArray()->all();
        $this->assertEquals(require(__DIR__ . '/data/test-find-galleries.php'), $data);
    }

    public function testFindGallery()
    {
        $gallery = Gallery::findOne(1);
        $this->assertEquals('image-1.jpg', $gallery->image1);
        $this->assertEquals('image-2.jpg', $gallery->image2);
    }

    public function testGetUploadWithMultiplyBehaviorsAttachedOne()
    {
        $gallery = Gallery::findOne(3);

        $gallery->setScenario('update');

        $this->assertTrue($gallery->save());

        $path1 = $gallery->getUploadPath('image1');
        $this->assertNull($path1);

        $url1 = $gallery->getThumbUploadUrl('image1', 'thumb');
        $this->assertContains('assets', $url1);

        $path2 = $gallery->getUploadPath('image2');
        $this->assertTrue(is_file($path2));
        $this->assertEquals(sha1_file($path2), sha1_file(__DIR__ . '/data/test-image.jpg'));

        $url2 = $gallery->getThumbUploadUrl('image2', 'thumb');
        $this->assertContains('upload/gallery', $url2);
    }

    public function testGetUploadWithMultiplyBehaviorsAttachedTwo()
    {
        $gallery = Gallery::findOne(4);

        $gallery->setScenario('update');

        $_FILES['Gallery[image1]'] = $_FILES['Gallery[image2]'];
        unset($_FILES['Gallery[image2]']);

        $this->assertTrue($gallery->save());

        $url1 = $gallery->getThumbUploadUrl('image1', 'thumb');
        $this->assertContains('upload/gallery', $url1);

        $path1 = $gallery->getUploadPath('image1');
        $this->assertTrue(is_file($path1));
        $this->assertEquals(sha1_file($path1), sha1_file(__DIR__ . '/data/test-image.jpg'));

        $path2 = $gallery->getUploadPath('image2');
        $this->assertNull($path2);

        $url2 = $gallery->getThumbUploadUrl('image2', 'thumb');
        $this->assertContains('assets', $url2);
    }

    public function testDeleteImage()
    {
        $gallery = Gallery::findOne(4);

        $gallery->setScenario('update');

        $this->assertTrue($gallery->save());

        $path = $gallery->getUploadPath('image2');
        $this->assertFileExists($path);

        $thumb = $gallery->getThumbUploadPath('image2', 'thumb');
        $preview = $gallery->getThumbUploadPath('image2', 'preview');

        $this->assertFileExists($thumb);
        $this->assertFileExists($preview);

        $gallery->deleteImage('image2');

        $this->assertEquals('', $gallery->image2);

        $this->assertFileNotExists($path);
        $this->assertFileNotExists($thumb);
        $this->assertFileNotExists($preview);
    }

    /**
     * @inheritdoc
     */
    public static function setUpBeforeClass()
    {
        parent::setUpBeforeClass();

        $_FILES = [
            'User[image]' => [
                'name' => 'test-image.jpg',
                'type' => 'image/jpeg',
                'size' => 74463,
                'tmp_name' => __DIR__ . '/fixtures/test-image.jpg',
                'error' => 0,
            ],
            'Gallery[image2]' => [
                'name' => 'test-image.jpg',
                'type' => 'image/jpeg',
                'size' => 74463,
                'tmp_name' => __DIR__ . '/fixtures/test-image.jpg',
                'error' => 0,
            ],
        ];
    }
}


src/UploadImageBehavior.php
<?php

namespace mohorev\file;

use Imagine\Image\ManipulatorInterface;
use Yii;
use yii\base\InvalidArgumentException;
use yii\base\InvalidConfigException;
use yii\base\NotSupportedException;
use yii\db\BaseActiveRecord;
use yii\helpers\ArrayHelper;
use yii\helpers\FileHelper;
use yii\imagine\Image;

/**
 * UploadImageBehavior automatically uploads image, creates thumbnails and fills
 * the specified attribute with a value of the name of the uploaded image.
 *
 * To use UploadImageBehavior, insert the following code to your ActiveRecord class:
 *
 * ```php
 * use mohorev\file\UploadImageBehavior;
 *
 * function behaviors()
 * {
 *     return [
 *         [
 *             'class' => UploadImageBehavior::class,
 *             'attribute' => 'file',
 *             'scenarios' => ['insert', 'update'],
 *             'placeholder' => '@app/modules/user/assets/images/userpic.jpg',
 *             'path' => '@webroot/upload/{id}/images',
 *             'url' => '@web/upload/{id}/images',
 *             'thumbPath' => '@webroot/upload/{id}/images/thumb',
 *             'thumbUrl' => '@web/upload/{id}/images/thumb',
 *             'createThumbsOnSave' => false,
 *             'createThumbsOnRequest' => true,
 *             'thumbs' => [
 *                   'thumb' => ['width' => 400, 'quality' => 90],
 *                   'preview' => ['width' => 200, 'height' => 200],
 *              ],
 *         ],
 *     ];
 * }
 * ```
 *
 * @author Alexander Mohorev <dev.mohorev@gmail.com>
 * @author Alexey Samoylov <alexey.samoylov@gmail.com>
 */
class UploadImageBehavior extends UploadBehavior
{
    /**
     * @var string
     */
    public $placeholder;
    /**
     * create all thumbs profiles on image upload
     * @var boolean
     */
    public $createThumbsOnSave = true;
    /**
     * create thumb only for profile request by getThumbUploadUrl() method
     * @var boolean
     */
    public $createThumbsOnRequest = false;
    /**
     * Whether delete original uploaded image after thumbs generating.
     * Defaults to FALSE
     * @var boolean
     */
    public $deleteOriginalFile = false;
    /**
     * @var array the thumbnail profiles
     * - `width`
     * - `height`
     * - `quality`
     */
    public $thumbs = [
        'thumb' => ['width' => 200, 'height' => 200, 'quality' => 90],
    ];
    /**
     * @var string|null
     */
    public $thumbPath;
    /**
     * @var string|null
     */
    public $thumbUrl;


    /**
     * @inheritdoc
     */
    public function init()
    {
        if (!class_exists(Image::class)) {
            throw new NotSupportedException("Yii2-imagine extension is required to use the UploadImageBehavior");
        }

        parent::init();

        if ($this->thumbPath === null) {
            $this->thumbPath = $this->path;
        }
        if ($this->thumbUrl === null) {
            $this->thumbUrl = $this->url;
        }

        foreach ($this->thumbs as $config) {
            $width = ArrayHelper::getValue($config, 'width');
            $height = ArrayHelper::getValue($config, 'height');
            if ($height < 1 && $width < 1) {
                throw new InvalidConfigException(sprintf(
                    'Length of either side of thumb cannot be 0 or negative, current size ' .
                    'is %sx%s', $width, $height
                ));
            }
        }
    }

    /**
     * @param string $attribute
     * @param string $profile
     * @return string|null
     * @throws \yii\base\Exception
     * @throws \yii\base\InvalidConfigException
     */
    public function getThumbUploadUrl($attribute, $profile = 'thumb')
    {
        /** @var BaseActiveRecord $model */
        $model = $this->owner;

        if ($this->attribute !== $attribute) {
            $behaviors = $model->getBehaviors();

            foreach ($behaviors as $behavior) {
                if ($behavior instanceof UploadImageBehavior) {
                    if ($behavior->attribute == $attribute) {
                        return $behavior->getThumbUploadUrl($attribute, $profile);
                    }
                }
            }
        }

        if (!$model->getAttribute($attribute)) {
            if ($this->placeholder) {
                return $this->getPlaceholderUrl($profile);
            } else {
                return null;
            }
        }

        $path = $this->getUploadPath($attribute, true);

        //if original file exist - generate profile thumb and generate url to thumb
        if (is_file($path) || !$this->deleteOriginalFile) {
            if ($this->createThumbsOnRequest) {
                $this->createThumbs($profile);
            }
            return $this->getThumbProfileUrl($attribute, $profile, $model);
        } //if original file is deleted generate url to thumb
        elseif ($this->deleteOriginalFile) {
            return $this->getThumbProfileUrl($attribute, $profile, $model);
        } elseif ($this->placeholder) {
            return $this->getPlaceholderUrl($profile);
        } else {
            return null;
        }
    }

    /**
     * @param $profile
     * @return string
     */
    protected function getPlaceholderUrl($profile)
    {
        list ($path, $url) = Yii::$app->assetManager->publish($this->placeholder);
        $filename = basename($path);
        $thumb = $this->getThumbFileName($filename, $profile);
        $thumbPath = dirname($path) . DIRECTORY_SEPARATOR . $thumb;
        $thumbUrl = dirname($url) . '/' . $thumb;

        if (!is_file($thumbPath)) {
            $this->generateImageThumb($this->thumbs[$profile], $path, $thumbPath);
        }

        return $thumbUrl;
    }

    /**
     * @param $attribute
     * @param $profile
     * @param BaseActiveRecord $model
     * @return bool|string
     */
    protected function getThumbProfileUrl($attribute, $profile, BaseActiveRecord $model)
    {
        $url = $this->resolvePath($this->thumbUrl);
        $fileName = $model->getOldAttribute($attribute);
        $thumbName = $this->getThumbFileName($fileName, $profile);

        return Yii::getAlias($url . '/' . $thumbName);
    }

    /**
     * @inheritdoc
     */
    protected function afterUpload()
    {
        parent::afterUpload();
        if ($this->createThumbsOnSave) {
            $this->createThumbs();
        }
    }

    /**
     * @param string $needed_profile - profile name to create thumb
     * @throws \yii\base\Exception
     * @throws \yii\base\InvalidConfigException
     */
    protected function createThumbs($needed_profile = false)
    {
        $path = $this->getUploadPath($this->attribute);
        foreach ($this->thumbs as $profile => $config) {
            //skip profiles not needed now
            if ($needed_profile && $needed_profile != $profile) {
                continue;
            }

            $thumbPath = $this->getThumbUploadPath($this->attribute, $profile);
            if ($thumbPath !== null) {
                if (!FileHelper::createDirectory(dirname($thumbPath))) {
                    throw new InvalidArgumentException(
                        "Directory specified in 'thumbPath' attribute doesn't exist or cannot be created."
                    );
                }
                if (!is_file($thumbPath) && !file_exists($thumbPath)) {
                    $this->generateImageThumb($config, $path, $thumbPath);
                }
            }
        }

        if ($this->deleteOriginalFile) {
            $this->deleteFile($path);
        }
    }

    /**
     * @param string $attribute
     * @param string $profile
     * @param boolean $old
     * @return string
     * @throws \yii\base\InvalidConfigException
     */
    public function getThumbUploadPath($attribute, $profile = 'thumb', $old = false)
    {
        /** @var BaseActiveRecord $model */
        $model = $this->owner;
        $path = $this->resolvePath($this->thumbPath);
        $attribute = ($old === true) ? $model->getOldAttribute($attribute) : $model->$attribute;
        $filename = $this->getThumbFileName($attribute, $profile);

        return $filename ? Yii::getAlias($path . '/' . $filename) : null;
    }

    /**
     * @param $filename
     * @param string $profile
     * @return string
     */
    protected function getThumbFileName($filename, $profile = 'thumb')
    {
        return $profile . '-' . $filename;
    }

    /**
     * @param $config
     * @param $path
     * @param $thumbPath
     */
    protected function generateImageThumb($config, $path, $thumbPath)
    {
        $width = ArrayHelper::getValue($config, 'width');
        $height = ArrayHelper::getValue($config, 'height');
        $quality = ArrayHelper::getValue($config, 'quality', 100);
        $mode = ArrayHelper::getValue($config, 'mode', ManipulatorInterface::THUMBNAIL_INSET);
        $bg_color = ArrayHelper::getValue($config, 'bg_color', 'FFF');

        if (!$width || !$height) {
            $image = Image::getImagine()->open($path);
            $ratio = $image->getSize()->getWidth() / $image->getSize()->getHeight();
            if ($width) {
                $height = ceil($width / $ratio);
            } else {
                $width = ceil($height * $ratio);
            }
        }

        // Fix error "PHP GD Allowed memory size exhausted".
        ini_set('memory_limit', '512M');
        //for big images size
        ini_set('max_execution_time', 60);
        Image::$thumbnailBackgroundColor = $bg_color;
        Image::thumbnail($path, $width, $height, $mode)->save($thumbPath, ['quality' => $quality]);
    }

    /**
     * @inheritdoc
     */
    protected function delete($attribute, $old = false)
    {
        $profiles = array_keys($this->thumbs);
        foreach ($profiles as $profile) {
            $path = $this->getThumbUploadPath($attribute, $profile, $old);
            $this->deleteFile($path);
        }
        parent::delete($attribute, $old);
    }

    /**
     * Remove image and all thumbs
     *
     * @param       $attribute
     * @param false $old
     */
    public function deleteImage($attribute, $old = false)
    {
        $this->delete($attribute, $old);
        $this->owner->updateAttributes([
            $attribute => ''
        ]);
    }
}
